File: Sources/MarketEvents/ReminderScheduler.swift
import Foundation
import DiscordBM
import MongoKitten
import Logging

// current schedule for each reporting hour (New York time)
//
//   before market open:
//     - 1pm the day before
//   after market close:
//     - 9am the day of
//     - 3pm the day of
//   during market hours:
//     - 1pm the day before
//     - 9am the day of
//

actor ReminderScheduler {
    private let reminderStore: ReminderStore
    private let discordClient: any DiscordClient
    private let logger: Logger
    
    private var isSending: Bool = false
    
    init(reminderStore: ReminderStore, discordClient: any DiscordClient, logger: Logger = Logger(label: "ReminderScheduler")) {
        self.reminderStore = reminderStore
        self.discordClient = discordClient
        self.logger = logger
    }
    
    func start() async throws {
        try await sendIfNeeded()
        
        // usually I would have a routine that does something like:
        //   check the upcoming reminders (from the database),
        //   calculate the instances based off of that,
        //   schedule a timer to fire approximately at that time
        // however the database contents can change, and calculating the instances is potentially expensive
        //   (mostly from needing to load the objects out of the database).
        // additionally, these reminders could be weeks apart - a timer may behave unexpectedly over such a long duration.
        //
        // instead, we'll simply schedule timers every day at the potential times we're interested in
        
        let wakeTimes = WakeTimes(start: .now)
        for date in wakeTimes {
            let waitInterval: TimeInterval = date.timeIntervalSinceNow
            
            let now: ContinuousClock.Instant = .now
            let target = now.advanced(by: Duration.seconds(waitInterval))
            try await Task.sleep(until: target)
            
            try await reminderStore.prune()
            
            try await sendIfNeeded()
        }
    }
    
    func sendIfNeeded() async throws {
        if isSending { return }
        
        isSending = true
        do {
            let stapledInstances = try await self.eligibleReminders(for: .now)
            await sendMessages(for: stapledInstances)
            isSending = false
        } catch {
            isSending = false
            throw error
        }
    }
    
    private func eligibleReminders(for date: Date) async throws -> [StapledReminderInstances] {
        let calendar = MarketEvents.newYorkPosixCalendar
        
        let startOfDay = calendar.startOfDay(for: date)
        // for simplicity, get the next 3 days
        guard let intervalEnd = calendar.date(byAdding: .day, value: 3, to: startOfDay) else {
            throw Calendar.ArithmeticError()
        }
        
        let dbEarningsDateKey = ReminderStore.Element.CodingKeys.earningsDate.stringValue
        let reminderCandidates = await reminderStore.find(
            matching: dbEarningsDateKey >= startOfDay && dbEarningsDateKey < intervalEnd
        )
        
        // to avoid requiring very precise timing, send reminders up to 100 seconds early
        let maxDate = date.addingTimeInterval(100)
        
        var collect: [StapledReminderInstances] = []
        for try await reminder in reminderCandidates {
            var stapled = reminder.stapledInstances(for: calendar)
            stapled.instances.removeAll { instance in
                reminder.sentKeys.contains(instance.key)
            }
            stapled.instances.removeAll { instance in
                instance.date > maxDate
            }
            
            if !stapled.instances.isEmpty {
                collect.append(stapled)
            }
        }
        return collect
    }
    
    private func sendMessages(for instances: [StapledReminderInstances]) async {
        let calendar = MarketEvents.newYorkPosixCalendar
        
        let channelBins: [ChannelSnowflake: [ReminderBin: [StapledReminderInstances]]] = instances.reduce(into: [:]) { partialResult, instance in
            let dateComponents = calendar.dateComponents([.year, .month, .day], from: instance.reminder.earningsDate)
            guard let year = dateComponents.year, let month = dateComponents.month, let day = dateComponents.day else {
                return // unexpected
            }
            let bin = ReminderBin(year: year, month: month, day: day, hour: instance.reminder.earningsHour)
            partialResult[instance.reminder.channelId, default: [:]][bin, default: []].append(instance)
        }
        
        await withTaskGroup(of: Void.self) { group in
            for (channelId, bins) in channelBins {
                group.addTask {
                    do {
                        try await self.sendMessage(to: channelId, bins: bins)
                    } catch {
                        self.logger.error("sendMessage(to:bins:)", metadata: [
                            "error": "\(error)"
                        ])
                    }
                }
            }
        }
    }
    
    private func sendMessage(to channelId: ChannelSnowflake, bins: [ReminderBin: [StapledReminderInstances]]) async throws {
        let datedSections: [(section: ReminderMessageSection, date: Date, hour: Finnhub.CheckedMarketHour)] = bins.compactMap { bin, instances in
            let section = ReminderMessageSection(instances: instances)
            guard let instance = instances.first else { return nil }
            return (section, instance.reminder.earningsDate, bin.hour)
        }
        
        let sortedSections = datedSections.sorted(using: [
            KeyPathComparator(\.date),
            KeyPathComparator(\.hour)
        ])
        
        let userFacingDateStyle = Date.FormatStyle(
            date: .long, time: .omitted, locale: MarketEvents.posixLocale,
            calendar: MarketEvents.newYorkPosixCalendar, timeZone: MarketEvents.newYorkTimeZone,
            capitalizationContext: .middleOfSentence
        )
        
        let symbolsFormat: ListFormatStyle<StringStyle, [String]> =
            .list(type: .and, width: .narrow)
            .locale(MarketEvents.posixLocale)
        
        let sectionContents: [String] = sortedSections.map { section, date, hour in
            var sectionMessage: String = "### "
            sectionMessage += date.formatted(userFacingDateStyle)
            switch hour {
            case .beforeMarketOpen:
                sectionMessage += " before market open"
            case .afterMarketClose:
                sectionMessage += " after market close"
            case .duringMarketHours:
                sectionMessage += " during market hours"
            }
            sectionMessage += "\n"
            
            let symbols = section.instances.map(\.reminder.symbol)
            sectionMessage += symbols.formatted(symbolsFormat)
            
            return sectionMessage
        }
        
        let message = Payloads.CreateMessage(embeds: [
            .init(title: "Earnings Reminders", description: sectionContents.joined(separator: "\n"))
        ])
        
        try await discordClient
            .createMessage(channelId: channelId, payload: message)
            .guardSuccess()
        
        let sentInstances = sortedSections.map(\.section).flatMap(\.instances)
        let taggedReminders: [ReminderStore.Element.ID: ReminderStore.Element] = sentInstances.reduce(into: [:]) { result, staple in
            let sentKeys = staple.instances.map(\.key)
            result[staple.reminder.id, default: staple.reminder].sentKeys.formUnion(sentKeys)
        }
        
        try await reminderStore.update(taggedReminders.values)
    }
}

private struct ReminderBin: Hashable {
    let year: Int
    let month: Int
    let day: Int
    
    let hour: Finnhub.CheckedMarketHour
}

private struct ReminderMessageSection {
    let instances: [StapledReminderInstances]
}

extension ReminderScheduler {
    nonisolated struct WakeTimes: Sequence, IteratorProtocol {
        let calendar = MarketEvents.newYorkPosixCalendar
        
        private var lookAhead: [Date] = []
        private var floor: Date
        
        init(start: Date) {
            self.floor = start
        }
        
        private func timesForDay(_ date: Date) -> [Date] {
            let startOfDay = calendar.startOfDay(for: date)
            
            let candidates: [Date?] = [
                // 9am
                calendar.date(bySettingHour: 9, minute: 0, second: 0, of: startOfDay),
                // 1pm
                calendar.date(bySettingHour: 13, minute: 0, second: 0, of: startOfDay),
                // 3pm
                calendar.date(bySettingHour: 15, minute: 0, second: 0, of: startOfDay),
            ]
            return candidates.compactMap(\.self)
        }
        
        mutating func next() -> Date? {
            if !lookAhead.isEmpty {
                let pop = lookAhead.removeFirst()
                floor = pop
                return pop
            }
            
            let todayCheck = timesForDay(floor)
                .drop { $0 < floor }
            
            if todayCheck.isEmpty {
                // start of day to avoid drift
                let startOfDay = calendar.startOfDay(for: floor)
                guard let tomorrow = calendar.date(byAdding: .day, value: 1, to: startOfDay) else {
                    return nil
                }
                floor = tomorrow
                return next()
            }
            
            lookAhead = Array(todayCheck)
            return next()
        }
    }
}


struct ReminderInstance {
    let key: String
    let date: Date
}

struct StapledReminderInstances {
    let reminder: ReminderStore.Element
    
    var instances: [ReminderInstance]
}

extension ReminderStore.Element {
    func reminderInstances(for calendar: Calendar) -> [ReminderInstance] {
        // `earningsDate` is already probably startOfDay, but just to be sure
        let startOfEarningsDate = calendar.startOfDay(for: self.earningsDate)
        
        var build: [ReminderInstance] = []
        
        switch self.earningsHour {
        case .beforeMarketOpen:
            // 1pm the day before
            if let dayBefore = calendar.date(byAdding: .day, value: -1, to: startOfEarningsDate),
               let yesterdayAfternoon = calendar.date(bySettingHour: 13, minute: 0, second: 0, of: dayBefore) {
                build.append(ReminderInstance(key: "1pm-yesterday", date: yesterdayAfternoon))
            }
        case .afterMarketClose:
            // 9am the day of
            if let morning = calendar.date(bySettingHour: 9, minute: 0, second: 0, of: startOfEarningsDate) {
                build.append(ReminderInstance(key: "9am-today", date: morning))
            }
            // 3pm the day of
            if let afternoon = calendar.date(bySettingHour: 15, minute: 0, second: 0, of: startOfEarningsDate) {
                build.append(ReminderInstance(key: "3pm-today", date: afternoon))
            }
        case .duringMarketHours:
            // 1pm the day before
            if let dayBefore = calendar.date(byAdding: .day, value: -1, to: startOfEarningsDate),
               let yesterdayAfternoon = calendar.date(bySettingHour: 13, minute: 0, second: 0, of: dayBefore) {
                build.append(ReminderInstance(key: "1pm-yesterday", date: yesterdayAfternoon))
            }
            // 9am the day of
            if let morning = calendar.date(bySettingHour: 9, minute: 0, second: 0, of: startOfEarningsDate) {
                build.append(ReminderInstance(key: "9am-today", date: morning))
            }
        }
        
        return build
    }
    
    func stapledInstances(for calendar: Calendar) -> StapledReminderInstances {
        return .init(reminder: self, instances: self.reminderInstances(for: calendar))
    }
}

File: src/reminderScheduler.ts
import { addDays, formatLongDateNewYork, formatSymbolList, newYorkYearMonthDay, setHourNewYork, startOfDayNewYork } from './dates';
import { createMessage } from './discord';
import { type CheckedMarketHour, MARKET_HOUR_ORDER } from './finnhub';
import type { Env } from './env';
import { type ReminderElement, ReminderStore } from './reminderStore';

// Schedule for each reporting hour (New York time):
//
//   before market open (bmo):
//     - 1pm the day before
//   after market close (amc):
//     - 9am the day of
//     - 3pm the day of
//   during market hours (dmh):
//     - 1pm the day before
//     - 9am the day of
//
// Replaces the Swift `ReminderScheduler` timer loop. Invoked by the cron trigger;
// `sent_keys` dedup makes it idempotent, so no in-flight lock is needed.

interface ReminderInstance {
	key: string;
	date: Date;
}

interface StapledReminder {
	reminder: ReminderElement;
	instances: ReminderInstance[];
}

export async function runScheduled(env: Env): Promise<void> {
	const reminderStore = new ReminderStore(env.DB);
	await reminderStore.prune();
	await sendIfNeeded(reminderStore, env, new Date());
}

function reminderInstances(reminder: ReminderElement): ReminderInstance[] {
	// `earnings_date` is already probably start-of-day, but just to be sure
	const startOfEarningsDate = startOfDayNewYork(reminder.earnings_date);
	const build: ReminderInstance[] = [];

	switch (reminder.earnings_hour) {
		case 'bmo':
			// 1pm the day before
			build.push({ key: '1pm-yesterday', date: setHourNewYork(addDays(startOfEarningsDate, -1), 13) });
			break;
		case 'amc':
			// 9am the day of
			build.push({ key: '9am-today', date: setHourNewYork(startOfEarningsDate, 9) });
			// 3pm the day of
			build.push({ key: '3pm-today', date: setHourNewYork(startOfEarningsDate, 15) });
			break;
		case 'dmh':
			// 1pm the day before
			build.push({ key: '1pm-yesterday', date: setHourNewYork(addDays(startOfEarningsDate, -1), 13) });
			// 9am the day of
			build.push({ key: '9am-today', date: setHourNewYork(startOfEarningsDate, 9) });
			break;
	}

	return build;
}

async function sendIfNeeded(reminderStore: ReminderStore, env: Env, now: Date): Promise<void> {
	const startOfDay = startOfDayNewYork(now);
	// for simplicity, get the next 3 days
	const intervalEnd = addDays(startOfDay, 3);

	// to avoid requiring very precise timing, send reminders up to 100 seconds early
	const maxDate = now.getTime() + 100 * 1000;

	const collected: StapledReminder[] = [];
	for (const reminder of await reminderStore.findInRange(startOfDay, intervalEnd)) {
		const sentKeys = new Set(reminder.sent_keys ?? []);
		const instances = reminderInstances(reminder).filter(
			(instance) => !sentKeys.has(instance.key) && instance.date.getTime() <= maxDate,
		);
		if (instances.length > 0) {
			collected.push({ reminder, instances });
		}
	}

	await sendMessages(reminderStore, env, collected);
}

async function sendMessages(reminderStore: ReminderStore, env: Env, staples: StapledReminder[]): Promise<void> {
	// channel -> (year/month/day/hour bin) -> stapled reminders
	const channelBins = new Map<string, Map<string, StapledReminder[]>>();
	for (const staple of staples) {
		const { year, month, day } = newYorkYearMonthDay(staple.reminder.earnings_date);
		const binKey = `${year}-${month}-${day}-${staple.reminder.earnings_hour}`;

		const bins = channelBins.get(staple.reminder.channel_id) ?? new Map<string, StapledReminder[]>();
		const bin = bins.get(binKey) ?? [];
		bin.push(staple);
		bins.set(binKey, bin);
		channelBins.set(staple.reminder.channel_id, bins);
	}

	const results = await Promise.allSettled(
		[...channelBins].map(([channelId, bins]) => sendMessage(reminderStore, env, channelId, bins)),
	);
	for (const result of results) {
		if (result.status === 'rejected') {
			console.error('sendMessage', result.reason);
		}
	}
}

interface ReminderSection {
	staples: StapledReminder[];
	date: Date;
	hour: CheckedMarketHour;
}

async function sendMessage(
	reminderStore: ReminderStore,
	env: Env,
	channelId: string,
	bins: Map<string, StapledReminder[]>,
): Promise<void> {
	const sections: ReminderSection[] = [];
	for (const staples of bins.values()) {
		const first = staples.at(0);
		if (!first) {
			continue;
		}
		sections.push({ staples, date: first.reminder.earnings_date, hour: first.reminder.earnings_hour });
	}

	sections.sort((lhs, rhs) => {
		const dateDiff = lhs.date.getTime() - rhs.date.getTime();
		if (dateDiff !== 0) {
			return dateDiff;
		}
		return MARKET_HOUR_ORDER[lhs.hour] - MARKET_HOUR_ORDER[rhs.hour];
	});

	const sectionContents = sections.map((section) => {
		let sectionMessage = `### ${formatLongDateNewYork(section.date)}`;
		switch (section.hour) {
			case 'bmo':
				sectionMessage += ' before market open';
				break;
			case 'amc':
				sectionMessage += ' after market close';
				break;
			case 'dmh':
				sectionMessage += ' during market hours';
				break;
		}
		sectionMessage += '\n';
		sectionMessage += formatSymbolList(section.staples.map((staple) => staple.reminder.symbol));
		return sectionMessage;
	});

	await createMessage(env.DISCORD_BOT_TOKEN, channelId, {
		embeds: [{ title: 'Earnings Reminders', description: sectionContents.join('\n') }],
	});

	// record which instance keys we just sent, per reminder
	const sentKeysByReminder = new Map<string, { channelId: string; symbol: string; keys: Set<string> }>();
	for (const section of sections) {
		for (const staple of section.staples) {
			// snowflakes shouldn't contain `_`, so this composite id shouldn't collide
			const mapKey = `${staple.reminder.symbol}_${staple.reminder.channel_id}`;
			const entry = sentKeysByReminder.get(mapKey) ?? {
				channelId: staple.reminder.channel_id,
				symbol: staple.reminder.symbol,
				keys: new Set<string>(),
			};
			for (const instance of staple.instances) {
				entry.keys.add(instance.key);
			}
			sentKeysByReminder.set(mapKey, entry);
		}
	}

	await Promise.all(
		[...sentKeysByReminder.values()].map((entry) => reminderStore.markSent(entry.channelId, entry.symbol, [...entry.keys])),
	);
}
